public/client/TracyCallstack.cpp

```diff
@@ -90,9 +90,111 @@ extern "C" const char* ___tracy_demangle( const char* mangled )
 #endif
 #endif
 
+#if TRACY_HAS_CALLSTACK == 3
+#   define TRACY_USE_IMAGE_CACHE
+#   include <link.h>
+#endif
+
 namespace tracy
 {
 
+#ifdef TRACY_USE_IMAGE_CACHE
+// when we have access to dl_iterate_phdr(), we can build a cache of address ranges to image paths
+// so we can quickly determine which image an address falls into.
+// We refresh this cache only when we hit an address that doesn't fall into any known range.
+class ImageCache
+{
+public:
+    struct ImageEntry
+    {
+        ImageEntry( void* startAddress, void* endAddress, const char* name )
+        : m_startAddress( startAddress ), m_endAddress( endAddress ), m_name( name ) {}
+
+        void* m_startAddress;
+        void* m_endAddress;
+        const char* m_name;
+    };
+
+    ImageCache()
+    {
+        m_images = (tracy::FastVector<ImageEntry>*)tracy::tracy_malloc( sizeof( tracy::FastVector<ImageEntry> ) );
+        new(m_images) tracy::FastVector<ImageEntry>( 512 );
+
+        Refresh();
+    }
+    const ImageEntry* GetImageForAddress( void* address ) 
+    {
+        const ImageEntry* entry = GetImageEntryForAddress( address );
+        if( !entry )
+        {
+            Refresh();
```

Comment:
Is there a mechanism that will prevent constant reloads when a large set of unmapped addresses is queried?

Reply:
There isn't one atm. The expectation would be that it wouldn't actually be called that often in real scenarios. In my case with a few hundred dynamic loaded shared libs, and instrumenting all memory allocations with a callstack captures I only get it called ~5 times until it actually has the full map as symbol resolution is relatively slow in itself.
What would be your suggestion, a "min ms between refreshes" and either skip (and return unknown symbols for that period) or "stall" the resolution thread for a period or time?

Reply:
Stalling to allow batch processing of a larger set may be a solution. But I was more interested in whether you have considered this case. I don't think this is immediately actionable.